hypothesis-python/src/hypothesis/internal/constants_ast.py

```diff
@@ -100,45 +113,60 @@ def _module_ast(module: ModuleType) -> Optional[AST]:
     return tree
 
 
-def local_modules() -> tuple[ModuleType, ...]:
-    modules = []
-    for module in sys.modules.values():
-        if (
-            not hasattr(module, "__file__")
-            or module.__file__ is None
-            # Skip expensive path lookup for stdlib modules.
-            # This will cause false negatives if a user names their module the
-            # same as a stdlib module.
-            #
-            # sys.stdlib_module_names is new in 3.10
-            or (
-                sys.version_info >= (3, 10)
-                and module.__name__ in sys.stdlib_module_names
-            )
-            or ModuleLocation.from_path(module.__file__) is not ModuleLocation.LOCAL
-        ):
-            continue
-
-        modules.append(module)
-    return tuple(modules)
-
+@lru_cache(4096)
+def _is_local_module_file(path: str) -> bool:
+    from hypothesis.internal.scrutineer import ModuleLocation
 
-def local_constants():
-    constants = set()
-    for module in local_modules():
+    return (
+        # Skip expensive path lookup for stdlib modules.
+        # This will cause false negatives if a user names their module the
+        # same as a stdlib module.
+        #
+        # sys.stdlib_module_names is new in 3.10
+        not (sys.version_info >= (3, 10) and path in sys.stdlib_module_names)
+        and ModuleLocation.from_path(path) is ModuleLocation.LOCAL
         # normally, hypothesis is a third-party library and is not returned
         # by local_modules. However, if it is installed as an editable package
         # with pip install -e, then we will pick up on it. Just hardcode an
         # ignore here.
+        and not is_hypothesis_file(path)
+        # avoid collecting constants from test files
+        and not ("test" in (p := Path(path)).parts or "tests" in p.parts)
```

Comment:
Let's add `or p.stem.startswith("test_") or p.stem.endswith("_test")` to the checks.